comp/core/workloadmeta/collectors/internal/remote/processcollector/process_collector.go

```diff
@@ -152,17 +171,40 @@ func (s *streamHandler) NewClient(cc grpc.ClientConnInterface) remote.GrpcClient
 	return &client{cl: pbgo.NewProcessEntityStreamClient(cc), parentCollector: s}
 }
 
-func (s *streamHandler) HandleResponse(resp interface{}) ([]workloadmeta.CollectorEvent, error) {
+// fetchContainerID updates the PID to Container ID mapping if at least one event is missing container ID field
+func (s *streamHandler) fetchContainerID(setEvents []*pbgo.ProcessEventSet, store workloadmeta.Component) {
+	requireFetch := false
+
+	for _, event := range setEvents {
+		if event != nil && event.GetContainerID() == "" {
+			requireFetch = true
+		}
+	}
+
+	if requireFetch {
+		containerProvider := proccontainers.GetSharedContainerProvider(store)
+		_, _, pidToCid, err := containerProvider.GetContainers(cacheValidityNoRT, s.lastContainerRates)
```

Comment:
instead of using `GetContainers()` https://github.com/DataDog/datadog-agent/blob/02c9011e8dad4ba85c8f91eadc1c544eae47eaeb/pkg/process/util/containers/containers.go#L100

maybe we should create another method that doesn't require to keep ` s.lastContainerRates` and that also retrieve container stats.

Reply:
Something like 

```golang
func (p *containerProvider) GetProcessIDToContainerIDMappign(cacheValidity time.Duration) (map[int]string, error) {
	containersMetadata := p.metadataStore.ListContainersWithFilter(workloadmeta.GetRunningContainers)

	pidToCid := make(map[int]string)
	for _, container := range containersMetadata {
		var annotations map[string]string
		if pod, err := p.metadataStore.GetKubernetesPodForContainer(container.ID); err == nil {
			annotations = pod.Annotations
		}

		if p.filter != nil && p.filter.IsExcluded(annotations, container.Name, container.Image.Name, container.Labels[kubernetes.CriContainerNamespaceLabel]) {
			continue
		}

		if container.Runtime == workloadmeta.ContainerRuntimeGarden && len(container.CollectorTags) == 0 {
			log.Debugf("No tags found for garden container: %s, skipping", container.ID)
			continue
		}

		entityID := containers.BuildTaggerEntityName(container.ID)
	

		// Building PID to CID mapping
		pids, err := collector.GetPIDs(container.Namespace, container.ID, cacheValidity)
		if err == nil && pids != nil {
			for _, pid := range pids {
				pidToCid[pid] = container.ID
			}
		} else {
			log.Debugf("PIDs for: %+v not available, err: %v", container, err)
		}
	}

	return pidToCid, nil
}
```

we can do better by refactoring some common code copy from the `GetContainers` function